File: OrdenarArquivos/src/main/kotlin/com/fenix/ordenararquivos/model/firebase/Manga.java
package com.fenix.ordenararquivos.model.firebase;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

public class Manga {

    private String nome;
    private String volume;
    private String capitulo;
    private String arquivo;
    private String capitulos;
    private Integer quantidade;
    private List<Caminhos> caminhos;
    private String sincronizacao;

    public String getNome() {
        return nome;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }

    public String getVolume() {
        return volume;
    }

    public void setVolume(String volume) {
        this.volume = volume;
    }

    public String getCapitulo() {
        return capitulo;
    }

    public void setCapitulo(String capitulo) {
        this.capitulo = capitulo;
    }

    public String getArquivo() {
        return arquivo;
    }

    public void setArquivo(String arquivo) {
        this.arquivo = arquivo;
    }

    public String getCapitulos() {
        return capitulos;
    }

    public void setCapitulos(String capitulos) {
        this.capitulos = capitulos;
    }

    public Integer getQuantidade() {
        return quantidade;
    }

    public void setQuantidade(Integer quantidade) {
        this.quantidade = quantidade;
    }

    public List<Caminhos> getCaminhos() {
        return caminhos;
    }

    public void setCaminhos(List<Caminhos> caminhos) {
        this.caminhos = caminhos;
    }

    public String getSincronizacao() {
        return sincronizacao;
    }

    public void setSincronizacao(String sincronizacao) {
        this.sincronizacao = sincronizacao;
    }

    public Manga(String nome, String volume, String capitulo, String arquivo, String capitulos, Integer quantidade, List<Caminhos> caminhos, String sincronizacao) {
        this.nome = nome;
        this.volume = volume;
        this.capitulo = capitulo;
        this.arquivo = arquivo;
        this.capitulos = capitulos;
        this.quantidade = quantidade;
        this.caminhos = caminhos;
        this.sincronizacao = sincronizacao;
    }

    public Manga(com.fenix.ordenararquivos.model.Manga manga) {
        this.nome = manga.getNome();
        this.volume = manga.getVolume();
        this.capitulo = manga.getCapitulo();
        this.arquivo = manga.getArquivo();
        this.capitulos = manga.getCapitulos();
        this.quantidade = manga.getQuantidade();
        this.caminhos = manga.getCaminhos().isEmpty() ? new ArrayList<>() : manga.getCaminhos().parallelStream().map(Caminhos::new).collect(Collectors.toList());
    }

    public static com.fenix.ordenararquivos.model.Manga toManga(Long id, HashMap<String, ?> obj) {
        return new com.fenix.ordenararquivos.model.Manga(id, (String) obj.get("nome"), (String) obj.get("volume"), (String) obj.get("capitulo"), (String) obj.get("arquivo"), (String) obj.get("capitulos"), ((Long) obj.get("quantidade")).intValue(), LocalDateTime.now(), new ArrayList<>());
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Manga manga = (Manga) o;
        return Objects.equals(nome, manga.nome) && Objects.equals(volume, manga.volume) && Objects.equals(capitulo, manga.capitulo) && Objects.equals(arquivo, manga.arquivo) && Objects.equals(capitulos, manga.capitulos);
    }

    @Override
    public int hashCode() {
        return Objects.hash(nome, volume, capitulo, arquivo, capitulos);
    }
}

File: OrdenarArquivos/src/main/kotlin/com/fenix/ordenararquivos/model/entities/Manga.kt
package com.fenix.ordenararquivos.model.entities

import com.google.firebase.database.Exclude
import java.time.LocalDateTime

data class Manga(
    var id: Long = 0,
    var nome: String = "",
    var volume: String = "",
    var capitulo: String = "",
    var arquivo: String = "",
    var capitulos: String = "",
    var quantidade: Int = 0,
    @Exclude @set:Exclude @get:Exclude var atualizacao: LocalDateTime = LocalDateTime.now(),
    var caminhos: MutableList<Caminhos> = arrayListOf()
) {
    constructor(id: Long, nome: String, volume: String, capitulo: String, arquivo: String, quantidade: Int, capitulos: String, atualizacao: LocalDateTime) : this(id, nome, volume, capitulo, arquivo, capitulos, quantidade, atualizacao) { }

    fun addCaminhos(caminhos: Caminhos) {
        this.caminhos.add(caminhos)
    }

    fun merge(manga: Manga) {
        this.nome = manga.nome
        this.volume = manga.volume
        this.capitulo = manga.capitulo
        this.arquivo = manga.arquivo
        this.capitulos = manga.capitulos
        this.quantidade = manga.quantidade
        this.caminhos = manga.caminhos
    }

    override fun toString(): String {
        return ("Manga [id=" + id + ", nome=" + nome + ", volume=" + volume + ", capitulo=" + capitulo + ", arquivo=" + arquivo + "]")
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Manga

        if (nome != other.nome) return false
        if (volume != other.volume) return false
        if (capitulo != other.capitulo) return false
        if (arquivo != other.arquivo) return false
        if (capitulos != other.capitulos) return false

        return true
    }

    override fun hashCode(): Int {
        var result = nome.hashCode()
        result = 31 * result + volume.hashCode()
        result = 31 * result + capitulo.hashCode()
        result = 31 * result + arquivo.hashCode()
        result = 31 * result + capitulos.hashCode()
        return result
    }

    companion object {
        fun copy(manga: Manga): Manga {
            return Manga(
                id = manga.id,
                nome = manga.nome,
                volume = manga.volume,
                capitulo = manga.capitulo,
                arquivo = manga.arquivo,
                capitulos = manga.capitulos,
                quantidade = manga.quantidade,
                atualizacao = manga.atualizacao,
                caminhos = manga.caminhos.toMutableList()
            )
        }
    }

}